Added the functions in brick ops

aujjwal-redhat opened the pull request:
Added get_online_bricks_list and get_offline_bricks_list.

Fixes: #475

Signed-off-by: Ayush Ujjwal <aujjwal@redhat.com>

aujjwal-redhat commented:
tested on a dummy code:
```js
# disruptive;rep

from tests.d_parent_test import DParentTest

class TestCase(DParentTest):

    def run_test(self, redant):
        redant.get_online_bricks_list(self.vol_name, self.server_list[0])

        redant.get_offline_bricks_list(self.vol_name, self.server_list[0])
```

CLI output:

```js
[root@servera redant]# python3 core/redant_main.py -c ../Utilities/conf.yaml -ll D -t tests/functional/glusterd/test_brick_ops.py -sp
Vol status:
{'test_brick_ops-rep': {'nodeCount': '5', 'node': [{'hostname': '10.70.43.74', 'path': '/glusterfs/brick/test_brick_ops-rep-0', 'peerid': 'adf863d7-9a2a-45c3-86a9-4f520c7e7092', 'status': '1', 'port': '49202', 'ports': {'tcp': '49202', 'rdma': 'N/A'}, 'pid': '250875'}, {'hostname': '10.70.43.115', 'path': '/glusterfs/brick/test_brick_ops-rep-1', 'peerid': '0bd0d62f-6f95-477f-9106-cd1c620def21', 'status': '1', 'port': '49154', 'ports': {'tcp': '49154', 'rdma': 'N/A'}, 'pid': '472804'}, {'hostname': '10.70.43.74', 'path': '/glusterfs/brick/test_brick_ops-rep-2', 'peerid': 'adf863d7-9a2a-45c3-86a9-4f520c7e7092', 'status': '1', 'port': '49203', 'ports': {'tcp': '49203', 'rdma': 'N/A'}, 'pid': '250891'}, {'hostname': 'Self-heal Daemon', 'path': 'localhost', 'peerid': 'adf863d7-9a2a-45c3-86a9-4f520c7e7092', 'status': '1', 'port': 'N/A', 'ports': {'tcp': 'N/A', 'rdma': 'N/A'}, 'pid': '250909'}, {'hostname': 'Self-heal Daemon', 'path': 'serverb', 'peerid': '0bd0d62f-6f95-477f-9106-cd1c620def21', 'status': '1', 'port': 'N/A', 'ports': {'tcp': 'N/A', 'rdma': 'N/A'}, 'pid': '472828'}], 'tasks': None}}

Online brick list: 
['10.70.43.74:/glusterfs/brick/test_brick_ops-rep-0', '10.70.43.115:/glusterfs/brick/test_brick_ops-rep-1', '10.70.43.74:/glusterfs/brick/test_brick_ops-rep-2']


Offline brick list: 
[]


test_brick_ops-rep PASS

Summary:
```

srijan-sivakumar commented on common/ops/gluster_ops/brick_ops.py:
```diff
@@ -465,3 +465,75 @@ def get_all_bricks(self, volname: str, node: str) -> list:
         self.logger.error(f"Bricks not found in the volume {volname}")
         return None
+
+    def get_online_bricks_list(self, volname: str, node: str):
```
nit: ->  def get_online_bricks_list(self, volname: str, node: str) -> list:

aujjwal-redhat replied:
ack

srijan-sivakumar commented on common/ops/gluster_ops/brick_ops.py:
```diff
@@ -465,3 +465,75 @@ def get_all_bricks(self, volname: str, node: str) -> list:
+            for brick in volume_status['node']:
+                if 'status' in brick:
+                    if int(brick['status']) == 1:
+                        cmp_brick_path = (f"{brick['hostname']}:"
```
will the hostname be always an IP when returned by `get_volume_status` ? If not then we need to have a check here to convert the data to IP if it comes as hostname.

nik-redhat replied:
I think it could be better if we handle this at the framework level (if possible) because in that case, I think many functions don't do this check of hostname/IP.

nik-redhat replied:
Are we always gonna have IP address in the brick list? Is that handled at the framework level?

srijan-sivakumar replied:
i think you're referring to the rexe module...but rexe just gets a response and sends it back to the overlying function...in this scenario it'd be `get_volume_status`. I think we cna handle it inside `get_volume_status`. What do you say ?

nik-redhat replied:
I think as per the latest discussion, we can ignore the conversion.

nik-redhat commented:
Could you do a test for the offline brick list as well?
Just stop the volume, before calling the function, that should work.

srijan-sivakumar commented on common/ops/gluster_ops/brick_ops.py:
```diff
@@ -465,3 +465,75 @@ def get_all_bricks(self, volname: str, node: str) -> list:
+        all_bricks = self.get_all_bricks(volname, node)
+
+        volume_status = volume_status[volname]
+        if 'node' in volume_status:
```
We are using the status to determine which bricks are offline, but what if a node is offline...will we consider the bricks lying under that brick as offline ? ( Probably we should ). I see that we are getting all the bricks from `get_all_bricks`, maybe we cna add one level of check to see if any of the bricks present in our all_brick is not even present in the response got from the volume_status. Those bricks we can directly add to the offline_brick_list

srijan-sivakumar commented:
> Could you do a test for the offline brick list as well?
> Just stop the volume, before calling the function, that should work.

Also check for offline node scenario..According to logic, I feel we might skip the brick of a downed node from offline brick list.

nik-redhat commented:
> > Could you do a test for the offline brick list as well?
> > Just stop the volume, before calling the function, that should work.
> 
> Also check for offline node scenario..According to logic, I feel we might skip the brick of a downed node from offline brick list.

Ahh, right. And, here double-check the logic from glusto, they don't handle this and we get a `key error` if the node is down.

aujjwal-redhat commented:
> > > Could you do a test for the offline brick list as well?
> > > Just stop the volume, before calling the function, that should work.
> > 
> > 
> > Also check for offline node scenario..According to logic, I feel we might skip the brick of a downed node from offline brick list.
> 
> Ahh, right. And, here double-check the logic from glusto, they don't handle this and we get a `key error` if the node is down.

In glusto they are traversing the bricks from the brick_list:
```js
for brick in bricks_list:
        brick_node, brick_path = brick.split(":")
        try:
            status = int(volume_status[volname]
                         [brick_node][brick_path]['status'])
        except KeyError:
            continue
```

Here, we are going through the nodes in the vol status and hence, we will not get the keyError. Moreover, that's why I have added a check : `if node in vol_status['node']` as well to double check and then traversing the brick list and checking if the brick is a part of the brick list populated from get_all_bricks.

aujjwal-redhat commented:
> Could you do a test for the offline brick list as well?
> Just stop the volume, before calling the function, that should work.

I think so, after stopping the volume,  volume status might not work if I am not wrong. Do correct me : )

srijan-sivakumar commented:
Yes...but we would still get some response which would sat the said volume
is not started. We need to check if this is handled.

On Fri, May 28, 2021 at 9:48 AM Ayush Ujjwal ***@***.***>
wrote:

> Could you do a test for the offline brick list as well?
> Just stop the volume, before calling the function, that should work.
>
> I think so, after stopping the volume, volume status might not work if I
> am not wrong. Do correct me : )
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/srijan-sivakumar/redant/pull/476#issuecomment-850118728>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APUBJL5KG4XHGSYYHN6MTH3TP4KQBANCNFSM45UUTXBQ>
> .
>


-- 
Regards,
Srijan

srijan-sivakumar commented:
Ack. But we also need to see if the bricks which are not shown in the
volume_status are considered as offline bricks. I think the current logic
might not handle it.

On Fri, May 28, 2021 at 9:47 AM Ayush Ujjwal ***@***.***>
wrote:

> Could you do a test for the offline brick list as well?
> Just stop the volume, before calling the function, that should work.
>
> Also check for offline node scenario..According to logic, I feel we might
> skip the brick of a downed node from offline brick list.
>
> Ahh, right. And, here double-check the logic from glusto, they don't
> handle this and we get a key error if the node is down.
>
> In glusto they are traversing the bricks from the brick_list:
>
> for brick in bricks_list:
>         brick_node, brick_path = brick.split(":")
>         try:
>             status = int(volume_status[volname]
>                          [brick_node][brick_path]['status'])
>         except KeyError:
>             continue
>
> Here, we are going through the nodes in the vol status and hence, we will
> not get the keyError. Moreover, that's why I have added a check : if node
> in vol_status['node'] as well to double check and then traversing the
> brick list and checking if the brick is a part of the brick list populated
> from get_all_bricks.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/srijan-sivakumar/redant/pull/476#issuecomment-850118481>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APUBJLZMO45A4XRSYQPNEULTP4KNDANCNFSM45UUTXBQ>
> .
>


-- 
Regards,
Srijan

aujjwal-redhat commented:
> Ack. But we also need to see if the bricks which are not shown in the volume_status are considered as offline bricks. I think the current logic might not handle it.
> […](#)
> On Fri, May 28, 2021 at 9:47 AM Ayush Ujjwal ***@***.***> wrote: Could you do a test for the offline brick list as well? Just stop the volume, before calling the function, that should work. Also check for offline node scenario..According to logic, I feel we might skip the brick of a downed node from offline brick list. Ahh, right. And, here double-check the logic from glusto, they don't handle this and we get a key error if the node is down. In glusto they are traversing the bricks from the brick_list: for brick in bricks_list: brick_node, brick_path = brick.split(":") try: status = int(volume_status[volname] [brick_node][brick_path]['status']) except KeyError: continue Here, we are going through the nodes in the vol status and hence, we will not get the keyError. Moreover, that's why I have added a check : if node in vol_status['node'] as well to double check and then traversing the brick list and checking if the brick is a part of the brick list populated from get_all_bricks. — You are receiving this because your review was requested. Reply to this email directly, view it on GitHub <[#476 (comment)](https://github.com/srijan-sivakumar/redant/pull/476#issuecomment-850118481)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/APUBJLZMO45A4XRSYQPNEULTP4KNDANCNFSM45UUTXBQ> .
> -- Regards, Srijan

In the current logic the volstatus bricks are checked against the get_all_bricks

Thats why

> Ack. But we also need to see if the bricks which are not shown in the volume_status are considered as offline bricks. I think the current logic might not handle it.
> […](#)
> On Fri, May 28, 2021 at 9:47 AM Ayush Ujjwal ***@***.***> wrote: Could you do a test for the offline brick list as well? Just stop the volume, before calling the function, that should work. Also check for offline node scenario..According to logic, I feel we might skip the brick of a downed node from offline brick list. Ahh, right. And, here double-check the logic from glusto, they don't handle this and we get a key error if the node is down. In glusto they are traversing the bricks from the brick_list: for brick in bricks_list: brick_node, brick_path = brick.split(":") try: status = int(volume_status[volname] [brick_node][brick_path]['status']) except KeyError: continue Here, we are going through the nodes in the vol status and hence, we will not get the keyError. Moreover, that's why I have added a check : if node in vol_status['node'] as well to double check and then traversing the brick list and checking if the brick is a part of the brick list populated from get_all_bricks. — You are receiving this because your review was requested. Reply to this email directly, view it on GitHub <[#476 (comment)](https://github.com/srijan-sivakumar/redant/pull/476#issuecomment-850118481)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/APUBJLZMO45A4XRSYQPNEULTP4KNDANCNFSM45UUTXBQ> .
> -- Regards, Srijan

The bricks that are not in the volume status can be offline ? I actually don't know about that.

aujjwal-redhat commented:
If the node is down, then the bricks won't show for that node in vol status. Will they also be considered as offline?

srijan-sivakumar commented:
yes, let's consider it as offline. This will be the case most of the times,
say when one of the node is offline ( as we have rebooted it ).

On Fri, May 28, 2021 at 10:05 AM Ayush Ujjwal ***@***.***>
wrote:

> If the node is down, then the bricks won't show for that node in vol
> status. Will they also be considered as offline?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/srijan-sivakumar/redant/pull/476#issuecomment-850124048>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APUBJL23AUO4OE33J4OQUMDTP4MSPANCNFSM45UUTXBQ>
> .
>


-- 
Regards,
Srijan

aujjwal-redhat commented:
> yes, let's consider it as offline. This will be the case most of the times, say when one of the node is offline ( as we have rebooted it ).
> […](#)
> On Fri, May 28, 2021 at 10:05 AM Ayush Ujjwal ***@***.***> wrote: If the node is down, then the bricks won't show for that node in vol status. Will they also be considered as offline? — You are receiving this because your review was requested. Reply to this email directly, view it on GitHub <[#476 (comment)](https://github.com/srijan-sivakumar/redant/pull/476#issuecomment-850124048)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/APUBJL23AUO4OE33J4OQUMDTP4MSPANCNFSM45UUTXBQ> .
> -- Regards, Srijan

Ack

aujjwal-redhat commented:
@srijan-sivakumar I stopped glusterd in the test and checked if its bricks are getting added to the offline bricks list:

```js
(redantenv) [root@servera redant]# python3 core/redant_main.py -c ../Utilities/conf.yaml -ll D -t tests/functional/glusterd/test_brick_ops.py -sp
Vol status: 
{'test_brick_ops-rep': {'nodeCount': '5', 'node': [{'hostname': '10.70.43.74', 'path': '/glusterfs/brick/test_brick_ops-rep-0', 'peerid': 'adf863d7-9a2a-45c3-86a9-4f520c7e7092', 'status': '1', 'port': '49200', 'ports': {'tcp': '49200', 'rdma': 'N/A'}, 'pid': '278554'}, {'hostname': '10.70.43.115', 'path': '/glusterfs/brick/test_brick_ops-rep-1', 'peerid': '0bd0d62f-6f95-477f-9106-cd1c620def21', 'status': '1', 'port': '49152', 'ports': {'tcp': '49152', 'rdma': 'N/A'}, 'pid': '475557'}, {'hostname': '10.70.43.74', 'path': '/glusterfs/brick/test_brick_ops-rep-2', 'peerid': 'adf863d7-9a2a-45c3-86a9-4f520c7e7092', 'status': '1', 'port': '49201', 'ports': {'tcp': '49201', 'rdma': 'N/A'}, 'pid': '278570'}, {'hostname': 'Self-heal Daemon', 'path': 'localhost', 'peerid': 'adf863d7-9a2a-45c3-86a9-4f520c7e7092', 'status': '1', 'port': 'N/A', 'ports': {'tcp': 'N/A', 'rdma': 'N/A'}, 'pid': '278596'}, {'hostname': 'Self-heal Daemon', 'path': 'serverb', 'peerid': '0bd0d62f-6f95-477f-9106-cd1c620def21', 'status': '1', 'port': 'N/A', 'ports': {'tcp': 'N/A', 'rdma': 'N/A'}, 'pid': '475575'}], 'tasks': None}}

Online Brick list: 
['10.70.43.74:/glusterfs/brick/test_brick_ops-rep-0', '10.70.43.115:/glusterfs/brick/test_brick_ops-rep-1', '10.70.43.74:/glusterfs/brick/test_brick_ops-rep-2']

Stopped glusterd on 10.70.43.115

Vol status after stop glusterd: 
{'test_brick_ops-rep': {'nodeCount': '3', 'node': [{'hostname': '10.70.43.74', 'path': '/glusterfs/brick/test_brick_ops-rep-0', 'peerid': 'adf863d7-9a2a-45c3-86a9-4f520c7e7092', 'status': '1', 'port': '49200', 'ports': {'tcp': '49200', 'rdma': 'N/A'}, 'pid': '278554'}, {'hostname': '10.70.43.74', 'path': '/glusterfs/brick/test_brick_ops-rep-2', 'peerid': 'adf863d7-9a2a-45c3-86a9-4f520c7e7092', 'status': '1', 'port': '49201', 'ports': {'tcp': '49201', 'rdma': 'N/A'}, 'pid': '278570'}, {'hostname': 'Self-heal Daemon', 'path': 'localhost', 'peerid': 'adf863d7-9a2a-45c3-86a9-4f520c7e7092', 'status': '1', 'port': 'N/A', 'ports': {'tcp': 'N/A', 'rdma': 'N/A'}, 'pid': '278596'}], 'tasks': None}}

Offline Bricks: 
['10.70.43.115:/glusterfs/brick/test_brick_ops-rep-1']

test_brick_ops-rep PASS
```

nik-redhat commented:
For, volume stop?

aujjwal-redhat commented:
> For, volume stop?

For volume stop, status failed.

```js
[2021-05-28 11:06:34,015] ERROR  - Volume test_brick_ops-rep is not started
[2021-05-28 11:06:34,016] ERROR  - Volume test_brick_ops-rep is not started
[2021-05-28 11:06:34,016] ERROR  - Traceback (most recent call last):
  File "./tests/d_parent_test.py", line 79, in parent_run_test
    self.run_test(self.redant)
  File "./tests/functional/glusterd/test_brick_ops.py", line 14, in run_test
    redant.get_offline_bricks_list(self.vol_name, self.server_list[0])
  File "./common/ops/gluster_ops/brick_ops.py", line 519, in get_offline_bricks_list
    volume_status = self.get_volume_status(volname, node)
  File "./common/ops/gluster_ops/volume_ops.py", line 541, in get_volume_status
    ret = self.execute_abstract_op_node(cmd, node)
  File "./common/ops/abstract_ops.py", line 49, in execute_abstract_op_node
    raise Exception(ret['msg']['opErrstr'])
Exception: Volume test_brick_ops-rep is not started
```

srijan-sivakumar commented:
> > For, volume stop?
> 
> For volume stop, status failed.
> 
> ```js
> [2021-05-28 11:06:34,015] ERROR  - Volume test_brick_ops-rep is not started
> [2021-05-28 11:06:34,016] ERROR  - Volume test_brick_ops-rep is not started
> [2021-05-28 11:06:34,016] ERROR  - Traceback (most recent call last):
>   File "./tests/d_parent_test.py", line 79, in parent_run_test
>     self.run_test(self.redant)
>   File "./tests/functional/glusterd/test_brick_ops.py", line 14, in run_test
>     redant.get_offline_bricks_list(self.vol_name, self.server_list[0])
>   File "./common/ops/gluster_ops/brick_ops.py", line 519, in get_offline_bricks_list
>     volume_status = self.get_volume_status(volname, node)
>   File "./common/ops/gluster_ops/volume_ops.py", line 541, in get_volume_status
>     ret = self.execute_abstract_op_node(cmd, node)
>   File "./common/ops/abstract_ops.py", line 49, in execute_abstract_op_node
>     raise Exception(ret['msg']['opErrstr'])
> Exception: Volume test_brick_ops-rep is not started
> ```

We have to handle this exception. When a volume is stopped, all bricks are offline.

aujjwal-redhat commented:
> > > For, volume stop?
> > 
> > 
> > For volume stop, status failed.
> > ```js
> > [2021-05-28 11:06:34,015] ERROR  - Volume test_brick_ops-rep is not started
> > [2021-05-28 11:06:34,016] ERROR  - Volume test_brick_ops-rep is not started
> > [2021-05-28 11:06:34,016] ERROR  - Traceback (most recent call last):
> >   File "./tests/d_parent_test.py", line 79, in parent_run_test
> >     self.run_test(self.redant)
> >   File "./tests/functional/glusterd/test_brick_ops.py", line 14, in run_test
> >     redant.get_offline_bricks_list(self.vol_name, self.server_list[0])
> >   File "./common/ops/gluster_ops/brick_ops.py", line 519, in get_offline_bricks_list
> >     volume_status = self.get_volume_status(volname, node)
> >   File "./common/ops/gluster_ops/volume_ops.py", line 541, in get_volume_status
> >     ret = self.execute_abstract_op_node(cmd, node)
> >   File "./common/ops/abstract_ops.py", line 49, in execute_abstract_op_node
> >     raise Exception(ret['msg']['opErrstr'])
> > Exception: Volume test_brick_ops-rep is not started
> > ```
> 
> We have to handle this exception. When a volume is stopped, all bricks are offline.

Yeah.

aujjwal-redhat commented:
```js
(redantenv) [root@servera redant]# python3 core/redant_main.py -c ../Utilities/conf.yaml -ll D -t tests/functional/glusterd/test_brick_ops.py -sp
Online:
['10.70.43.74:/glusterfs/brick/test_brick_ops-rep-0', '10.70.43.115:/glusterfs/brick/test_brick_ops-rep-1', '10.70.43.74:/glusterfs/brick/test_brick_ops-rep-2']

Stopped glusterd on 10.70.43.115

Offline:
['10.70.43.115:/glusterfs/brick/test_brick_ops-rep-1']

Started glusterd on 10.70.43.115

Online:
['10.70.43.74:/glusterfs/brick/test_brick_ops-rep-0', '10.70.43.115:/glusterfs/brick/test_brick_ops-rep-1', '10.70.43.74:/glusterfs/brick/test_brick_ops-rep-2']

Stopped volume test_brick_ops-rep

Offline:
['10.70.43.74:/glusterfs/brick/test_brick_ops-rep-0', '10.70.43.115:/glusterfs/brick/test_brick_ops-rep-1', '10.70.43.74:/glusterfs/brick/test_brick_ops-rep-2']
test_brick_ops-rep PASS
```

aujjwal-redhat commented:
```js
(redantenv) [root@servera redant]# python3 core/redant_main.py -c ../Utilities/conf.yaml -ll D -t tests/functional/glusterd/test_brick_ops.py -sp

Online:
['10.70.43.74:/glusterfs/brick/test_brick_ops-rep-0', '10.70.43.115:/glusterfs/brick/test_brick_ops-rep-1', '10.70.43.74:/glusterfs/brick/test_brick_ops-rep-2']

Stopped glusterd on 10.70.43.115

Online:
['10.70.43.74:/glusterfs/brick/test_brick_ops-rep-0', '10.70.43.74:/glusterfs/brick/test_brick_ops-rep-2']

Offline:
['10.70.43.115:/glusterfs/brick/test_brick_ops-rep-1']

Started glusterd on 10.70.43.115

Online:
['10.70.43.74:/glusterfs/brick/test_brick_ops-rep-0', '10.70.43.115:/glusterfs/brick/test_brick_ops-rep-1', '10.70.43.74:/glusterfs/brick/test_brick_ops-rep-2']

Offline:
[]

Stopped volume test_brick_ops-rep

Online:
None

Offline:
['10.70.43.74:/glusterfs/brick/test_brick_ops-rep-0', '10.70.43.115:/glusterfs/brick/test_brick_ops-rep-1', '10.70.43.74:/glusterfs/brick/test_brick_ops-rep-2']

aujjwal-redhat commented:
@srijan-sivakumar @nik-redhat are all the cases covered up?

srijan-sivakumar commented:
@aujjwal-redhat let's create a example test code for showcasing this feature. We can keep it for reference if we need to understand how this feature works. We can add it inside the example/sample_component directory in tests

srijan-sivakumar commented:
> @srijan-sivakumar @nik-redhat are all the cases covered up?

Looks that way. I think we are good to go. Just a small example test would suffice.

nik-redhat commented:
> > @srijan-sivakumar @nik-redhat are all the cases covered up?
> 
> Looks that way. I think we are good to go. Just a small example test would suffice.

@srijan-sivakumar let's merge this, and take that up in a separate patch?

srijan-sivakumar commented:
> > > @srijan-sivakumar @nik-redhat are all the cases covered up?
> > 
> > 
> > Looks that way. I think we are good to go. Just a small example test would suffice.
> 
> @srijan-sivakumar let's merge this, and take that up in a separate patch?

cool. That sounds good to me.

aujjwal-redhat commented:
> > @srijan-sivakumar @nik-redhat are all the cases covered up?
> 
> Looks that way. I think we are good to go. Just a small example test would suffice.

@srijan-sivakumar @nik-redhat I have added the test case in [this patch](https://github.com/srijan-sivakumar/redant/pull/490)
